File: Android/TUIKit/TUICore/tuicore/src/main/java/com/tencent/qcloud/tuicore/permission/PermissionActivity.java
package com.tencent.qcloud.tuicore.permission;

import static com.tencent.qcloud.tuicore.permission.PermissionRequester.PERMISSION_NOTIFY_EVENT_KEY;
import static com.tencent.qcloud.tuicore.permission.PermissionRequester.PERMISSION_NOTIFY_EVENT_SUB_KEY;
import static com.tencent.qcloud.tuicore.permission.PermissionRequester.PERMISSION_RESULT;

import android.annotation.SuppressLint;
import android.app.ActionBar;
import android.app.Activity;
import android.app.AlertDialog;
import android.app.Dialog;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.provider.Settings;
import android.util.Log;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.RequiresApi;

import com.tencent.qcloud.tuicore.R;
import com.tencent.qcloud.tuicore.TUICore;
import com.tencent.qcloud.tuicore.util.TUIBuild;

import java.util.HashMap;
import java.util.Map;

@RequiresApi(api = Build.VERSION_CODES.M)
public class PermissionActivity extends Activity {
    private static final String TAG = "PermissionActivity";
    private static final int PERMISSION_REQUEST_CODE = 100;

    private TextView mRationaleTitleTv;
    private TextView mRationaleDescriptionTv;
    private ImageView mPermissionIconIv;
    private PermissionRequester.RequestData mRequestData;

    private PermissionRequester.Result mResult = PermissionRequester.Result.Denied;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        mRequestData = getPermissionRequest();
        if (mRequestData == null || mRequestData.isPermissionsExistEmpty()) {
            Log.e(TAG, "onCreate mRequestData exist empty permission");
            finishWithResult(PermissionRequester.Result.Denied);
            return;
        }
        Log.i(TAG, "onCreate : " + mRequestData.toString());
        if (TUIBuild.getVersionInt() < Build.VERSION_CODES.M) {
            finishWithResult(PermissionRequester.Result.Granted);
            return;
        }

        makeBackGroundTransparent();
        initView();
        showPermissionRationale();

        requestPermissions(mRequestData.getPermissions(), PERMISSION_REQUEST_CODE);
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);

        hidePermissionRationale();
        if (isAllPermissionsGranted(grantResults)) {
            finishWithResult(PermissionRequester.Result.Granted);
            return;
        }
        showSettingsTip();
    }

    private void showSettingsTip() {
        if (mRequestData == null) {
            return;
        }

        View tipLayout = LayoutInflater.from(this).inflate(R.layout.permission_tip_layout, null);
        TextView tipsTv = tipLayout.findViewById(R.id.tips);
        TextView positiveBtn = tipLayout.findViewById(R.id.positive_btn);
        TextView negativeBtn = tipLayout.findViewById(R.id.negative_btn);
        tipsTv.setText(mRequestData.getSettingsTip());

        Dialog permissionTipDialog = new AlertDialog.Builder(this).setView(tipLayout).setCancelable(false).create();

        positiveBtn.setOnClickListener(v -> {
            permissionTipDialog.dismiss();
            launchAppDetailsSettings();
            finishWithResult(PermissionRequester.Result.Requesting);
        });

        negativeBtn.setOnClickListener(v -> {
            permissionTipDialog.dismiss();
            finishWithResult(PermissionRequester.Result.Denied);
        });
        permissionTipDialog.setOnKeyListener((dialog, keyCode, event) -> {
            if (keyCode == KeyEvent.KEYCODE_BACK && event.getAction() == KeyEvent.ACTION_UP) {
                permissionTipDialog.dismiss();
            }
            return true;
        });

        Window dialogWindow = permissionTipDialog.getWindow();
        dialogWindow.setBackgroundDrawable(new ColorDrawable());
        WindowManager.LayoutParams layoutParams = dialogWindow.getAttributes();
        layoutParams.width = WindowManager.LayoutParams.WRAP_CONTENT;
        layoutParams.height = WindowManager.LayoutParams.WRAP_CONTENT;
        dialogWindow.setAttributes(layoutParams);
        permissionTipDialog.show();
    }

    /**
     * 启动应用程序的详细信息设置。
     *
     * Launch the application's details settings.
     */
    private void launchAppDetailsSettings() {
        Intent intent = new Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS);
        intent.setData(Uri.parse("package:" + getPackageName()));
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        if (getPackageManager().queryIntentActivities(intent, PackageManager.MATCH_DEFAULT_ONLY).size() <= 0) {
            Log.e(TAG, "launchAppDetailsSettings can not find system settings");
            return;
        }
        startActivity(intent);
    }

    private void finishWithResult(PermissionRequester.Result result) {
        Log.i(TAG, "finishWithResult : " + result);
        mResult = result;
        finish();
    }

    /*
     * 1、连续申请权限时，需要前一个权限申请完全结束；
     * 2、APP 从异常中恢复时，有些 Activity 在 onCreate 时就会申请权限；同时对于异常恢复，所有 Activity 可能被干掉，从新走流程；
     * 所以可能导致本次权限申请过早结束，此时须在 onDestroy 中返回结果。
     */
    @Override
    protected void onDestroy() {
        super.onDestroy();
        Map<String, Object> result = new HashMap<>(1);
        result.put(PERMISSION_RESULT, mResult);
        TUICore.notifyEvent(PERMISSION_NOTIFY_EVENT_KEY, PERMISSION_NOTIFY_EVENT_SUB_KEY, result);
    }

    private PermissionRequester.RequestData getPermissionRequest() {
        Intent intent = getIntent();
        if (intent == null) {
            return null;
        }
        return intent.getParcelableExtra(PermissionRequester.PERMISSION_REQUEST_KEY);
    }

    @SuppressLint("NewApi")
    private void makeBackGroundTransparent() {
        if (TUIBuild.getVersionInt() >= Build.VERSION_CODES.LOLLIPOP) {
            View decorView = getWindow().getDecorView();
            decorView.setSystemUiVisibility(View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN | View.SYSTEM_UI_FLAG_LAYOUT_STABLE);
            getWindow().setStatusBarColor(Color.TRANSPARENT);
            getWindow().setNavigationBarColor(Color.TRANSPARENT);
        }

        ActionBar actionBar = getActionBar();
        if (actionBar != null) {
            actionBar.hide();
        }
    }

    private void initView() {
        setContentView(R.layout.permission_activity_layout);
        mRationaleTitleTv = findViewById(R.id.permission_reason_title);
        mRationaleDescriptionTv = findViewById(R.id.permission_reason);
        mPermissionIconIv = findViewById(R.id.permission_icon);
    }

    /**
     * 安全合规要求，申请权限时，必须显示申请权限的理由。
     *
     * Security compliance requires that when applying for permission, the reason for applying for permission must be
     * displayed.
     */
    private void showPermissionRationale() {
        if (mRequestData == null) {
            return;
        }
        mRationaleTitleTv.setText(mRequestData.getTitle());
        mRationaleDescriptionTv.setText(mRequestData.getDescription());
        mPermissionIconIv.setBackgroundResource(mRequestData.getPermissionIconId());

        mRationaleTitleTv.setVisibility(View.VISIBLE);
        mRationaleDescriptionTv.setVisibility(View.VISIBLE);
        mPermissionIconIv.setVisibility(View.VISIBLE);
    }

    private void hidePermissionRationale() {
        mRationaleTitleTv.setVisibility(View.INVISIBLE);
        mRationaleDescriptionTv.setVisibility(View.INVISIBLE);
        mPermissionIconIv.setVisibility(View.INVISIBLE);
    }

    private boolean isAllPermissionsGranted(@NonNull int[] grantResults) {
        for (int result : grantResults) {
            if (result != PackageManager.PERMISSION_GRANTED) {
                return false;
            }
        }
        return true;
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        return true;
    }
}

File: Android/TUIKit/atomic_x/src/main/java/io/trtc/tuikit/atomicx/common/permission/PermissionActivity.kt
package io.trtc.tuikit.atomicx.common.permission

import android.annotation.SuppressLint
import android.app.Activity
import android.app.AlertDialog
import android.app.Dialog
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.os.Build
import android.os.Bundle
import android.provider.Settings
import android.util.Log
import android.view.KeyEvent
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.View
import android.view.WindowManager
import android.widget.ImageView
import android.widget.TextView
import androidx.annotation.RequiresApi
import androidx.core.net.toUri
import io.trtc.tuikit.atomicx.R
import io.trtc.tuikit.atomicx.common.event.EventManager
import io.trtc.tuikit.atomicx.common.permission.PermissionRequester.Companion.PERMISSION_NOTIFY_EVENT_KEY
import io.trtc.tuikit.atomicx.common.permission.PermissionRequester.Companion.PERMISSION_NOTIFY_EVENT_SUB_KEY
import io.trtc.tuikit.atomicx.common.permission.PermissionRequester.Companion.PERMISSION_REQUEST_KEY
import io.trtc.tuikit.atomicx.common.permission.PermissionRequester.Companion.PERMISSION_RESULT
import io.trtc.tuikit.atomicx.common.util.TUIBuild

@RequiresApi(api = Build.VERSION_CODES.M)
class PermissionActivity : Activity() {

    companion object {
        private const val TAG = "PermissionActivity"
        private const val PERMISSION_REQUEST_CODE = 100
    }

    private lateinit var rationaleTitleTv: TextView
    private lateinit var rationaleDescriptionTv: TextView
    private lateinit var permissionIconIv: ImageView
    private var requestData: PermissionRequester.RequestData? = null

    private var result = PermissionRequester.Result.Denied

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        requestData = getPermissionRequest()
        if (requestData == null || requestData?.isPermissionsExistEmpty() == true) {
            Log.e(TAG, "onCreate mRequestData exist empty permission")
            finishWithResult(PermissionRequester.Result.Denied)
            return
        }
        Log.i(TAG, "onCreate : ${requestData.toString()}")
        if (TUIBuild.getVersionInt() < Build.VERSION_CODES.M) {
            finishWithResult(PermissionRequester.Result.Granted)
            return
        }

        makeBackGroundTransparent()
        initView()
        showPermissionRationale()

        requestPermissions(
            (requestData?.getPermissions() ?: emptyArray()) as Array<String>,
            PERMISSION_REQUEST_CODE
        )
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)

        hidePermissionRationale()
        if (isAllPermissionsGranted(grantResults)) {
            finishWithResult(PermissionRequester.Result.Granted)
            return
        }
        showSettingsTip()
    }

    private fun showSettingsTip() {
        if (requestData == null) {
            return
        }

        val tipLayout =
            LayoutInflater.from(this).inflate(R.layout.common_permission_tip_layout, null)
        val tipsTv = tipLayout.findViewById<TextView>(R.id.tips)
        val positiveBtn = tipLayout.findViewById<TextView>(R.id.positive_btn)
        val negativeBtn = tipLayout.findViewById<TextView>(R.id.negative_btn)
        tipsTv.text = requestData?.settingsTip

        val permissionTipDialog: Dialog = AlertDialog.Builder(this)
            .setView(tipLayout)
            .setCancelable(false)
            .create()

        positiveBtn.setOnClickListener {
            permissionTipDialog.dismiss()
            launchAppDetailsSettings()
            finishWithResult(PermissionRequester.Result.Requesting)
        }

        negativeBtn.setOnClickListener {
            permissionTipDialog.dismiss()
            finishWithResult(PermissionRequester.Result.Denied)
        }

        permissionTipDialog.setOnKeyListener { dialog, keyCode, event ->
            if (keyCode == KeyEvent.KEYCODE_BACK && event.action == KeyEvent.ACTION_UP) {
                permissionTipDialog.dismiss()
            }
            true
        }

        val dialogWindow = permissionTipDialog.window
        dialogWindow?.setBackgroundDrawable(ColorDrawable())
        val layoutParams = dialogWindow?.attributes
        layoutParams?.width = WindowManager.LayoutParams.WRAP_CONTENT
        layoutParams?.height = WindowManager.LayoutParams.WRAP_CONTENT
        dialogWindow?.attributes = layoutParams
        permissionTipDialog.show()
    }

    /**
     * Launch the application's details settings.
     */
    private fun launchAppDetailsSettings() {
        val intent = Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS)
        intent.data = "package:$packageName".toUri()
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        if (packageManager.queryIntentActivities(
                intent,
                PackageManager.MATCH_DEFAULT_ONLY
            ).isEmpty()
        ) {
            Log.e(TAG, "launchAppDetailsSettings can not find system settings")
            return
        }
        startActivity(intent)
    }

    private fun finishWithResult(result: PermissionRequester.Result) {
        Log.i(TAG, "finishWithResult : $result")
        this@PermissionActivity.result = result
        finish()
    }

    override fun onDestroy() {
        super.onDestroy()
        val result = hashMapOf<String, Any>(PERMISSION_RESULT to result)
        EventManager.instance.notifyEvent(
            PERMISSION_NOTIFY_EVENT_KEY,
            PERMISSION_NOTIFY_EVENT_SUB_KEY,
            result
        )
    }

    private fun getPermissionRequest(): PermissionRequester.RequestData? {
        val intent = intent ?: return null
        return intent.getParcelableExtra(PERMISSION_REQUEST_KEY)
    }

    @SuppressLint("NewApi")
    private fun makeBackGroundTransparent() {
        if (TUIBuild.getVersionInt() >= Build.VERSION_CODES.LOLLIPOP) {
            val decorView = window.decorView
            decorView.systemUiVisibility =
                View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN or View.SYSTEM_UI_FLAG_LAYOUT_STABLE
            window.statusBarColor = Color.TRANSPARENT
            window.navigationBarColor = Color.TRANSPARENT
        }

        actionBar?.hide()
    }

    private fun initView() {
        setContentView(R.layout.common_permission_activity_layout)
        rationaleTitleTv = findViewById(R.id.permission_reason_title)
        rationaleDescriptionTv = findViewById(R.id.permission_reason)
        permissionIconIv = findViewById(R.id.permission_icon)
    }

    /**
     *
     * Security compliance requires that when applying for permission, the reason for applying for permission must be
     * displayed.
     */
    private fun showPermissionRationale() {
        if (requestData == null) {
            return
        }
        rationaleTitleTv.text = requestData?.title
        rationaleDescriptionTv.text = requestData?.description
        permissionIconIv.setBackgroundResource(requestData?.permissionIconId ?: 0)

        rationaleTitleTv.visibility = View.VISIBLE
        rationaleDescriptionTv.visibility = View.VISIBLE
        permissionIconIv.visibility = View.VISIBLE
    }

    private fun hidePermissionRationale() {
        rationaleTitleTv.visibility = View.INVISIBLE
        rationaleDescriptionTv.visibility = View.INVISIBLE
        permissionIconIv.visibility = View.INVISIBLE
    }

    private fun isAllPermissionsGranted(grantResults: IntArray): Boolean {
        for (result in grantResults) {
            if (result != PackageManager.PERMISSION_GRANTED) {
                return false
            }
        }
        return true
    }

    override fun onTouchEvent(event: MotionEvent?): Boolean {
        return true
    }
}
